Feature/passport visa

Azher2Ali opened the pull request:
This PR contains changes related to the below items :

1. Create a visa service that will : #703 

- parseVisa → Visa class that is returned from content of Visa JWT (string)
- validateVisa → need public key? or do we just trust the passport broker (i think we just trust the broker so these are already valid)
- Caching pubic key using Spring Cache (CacheUtil) fetched from passport broker getPublicKey endpoint (maintained at application.yaml) and evicting every midnight
- getPermissionsForVisa → return list of permissions for visa

2. Create a Passport Service : #701 

- parsePassport → return Passport object from content of Passport JWT (string)
- validatePassport (vs public key) → requires passportBroker config (to fetch public key) 
- Caching pubic key using Spring Cache (CacheUtil) fetched from passport broker getPublicKey endpoint (maintained at application.yaml) and evicting every midnight
- getVisas → return list of visa JWTs
- getPermissionsForPassport → return list of permissions from passport (compiled list of all permissions from all visas, remove duplicates)

![image](https://github.com/overture-stack/ego/assets/121898125/437de56d-1865-434f-9949-eec2fcf8c747)

Azher2Ali committed:
Creating Passport Service

Azher2Ali committed:
Committing changes related to visa service and passport service

Azher2Ali committed:
Committing changes related to visa service and passport service

leoraba commented on src/main/java/bio/overture/ego/service/PassportService.java:
```diff
@@ -0,0 +1,104 @@
+
+  // Parse Passport token to extract the passport body
+  public Passport parsePassport(@NonNull String passportJwtToken) throws JsonProcessingException {
+    String[] split_string = passportJwtToken.split("//.");
```
regex syntax uses backlash to scape special characters. in this case should be like "\\."

Azher2Ali replied:
Resolved

leoraba commented on src/main/java/bio/overture/ego/service/VisaService.java:
```diff
@@ -65,6 +66,18 @@ public void delete(@NonNull UUID id) {
 
+  // Parses Visa JWT token to convert into Visa Object
+  public Visa parseVisa(@NonNull String visaJwtToken) throws JsonProcessingException {
+    String[] split_string = visaJwtToken.split("//.");
```
regex syntax uses backlash to scape special characters. in this case should be like `\\.`

leoraba commented on src/main/java/bio/overture/ego/model/dto/Passport.java:
```diff
@@ -0,0 +1,40 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class Passport {
```
Passport could contain additional/optional fields as per [spec](https://ga4gh.github.io/data-security/aai-openid-connect-profile#passport-format)

consider adding mapping annotation to ignore unknown fields:
`@JsonIgnoreProperties(ignoreUnknown = true)`

Azher2Ali replied:
Added JsonIgnoreProperties

Azher2Ali committed:
Committing changes related to passport permissions

Azher2Ali committed:
Committing changes related to passport permissions

Azher2Ali committed:
Committing changes related to passport permissions

Azher2Ali committed:
Committing changes related to passport visa

Azher2Ali committed:
Committing changes related to passport visa

leoraba commented on src/main/resources/application.yml:
```diff
@@ -184,6 +184,12 @@ token:
+broker:
+  token:
+    issuer: http://localhost:8081
+    private-key: MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQDSU6oy48sJW6xzqzOSU1dAvUUeFKQSBHsCf7wGWUGpOxEczhtFiiyx4YUJtg+fyvwWxa4wO3GnQLBPIxBHY8JsnvjQN2lsTUoLqMB9nGpwF617uA/S2igm1u+cDpfi82kbi6SG1Sg30PM047R6oxTRGDLLkeMRF1gRaTBM0HfSL0j6ccU5KPgwYsFLE2We6jeR56iYJGC2KYLH4v8rcc2jRAdMbUntHMtUByF9BPSW7elQnyQH5Qzr/o0b59XLKwnJFn2Bp2yviC8cdyTDyhQGna0e+oESQR1j6u3Ux/mOmm3slRXscA8sH+pHmOEAtjYVf/ww36U8uZ ...
```
we shouldn't keep our private/public keys on git repo.

Azher2Ali replied:
The pubic/private key used here was a dummy one which was required during testing. As per the requested change, following keys are being removed from the git repo

Azher2Ali committed:
Update application.yml

The private/public key mentioned in the git repo was a dummy one, which was used during testing. As per the requested change, public/private keys are removed from the application.yaml

leoraba reviewed (approved):
Looks good to me. 👍

Azher2Ali committed:
Committing changes related to caching broker public key

Azher2Ali committed:
Merge remote-tracking branch 'origin/feature/passportVisa' into feature/passportVisa

# Conflicts:
#	src/main/resources/application.yml

leoraba reviewed (approved):
👍 approved!
